airflow/contrib/hooks/yandexcloud_dataproc_hook.py

```diff
@@ -0,0 +1,571 @@
+                        ):
+        """
+        Run Pyspark job in Yandex.Cloud Data Proc cluster.
+
+        :param main_python_file_uri: URI of python file with job. Can be placed in HDFS or S3.
+        :type main_python_file_uri: str
+        :param python_file_uris: URIs of python files used in the job. Can be placed in HDFS or S3.
+        :type python_file_uris: List[str]
+        :param file_uris: URIs of files used in the job. Can be placed in HDFS or S3.
+        :type file_uris: List[str]
+        :param archive_uris: URIs of archive files used in the job. Can be placed in HDFS or S3.
+        :type archive_uris: List[str]
+        :param jar_file_uris: URIs of JAR files used in the job. Can be placed in HDFS or S3.
+        :type archive_uris: List[str]
+        :param properties: Properties for the job.
+        :type properties: Dist[str, str]
+        :param arguments: Arguments to be passed to the job.
+        :type arguments: List[str]
+        :param cluster_id: ID of the cluster to run job in.
+                           Will try to take the ID from Dataproc Hook object if ot specified.
+        :type cluster_id: str
+        :param name: Name of the job. Used for labeling.
+        :type name: str
+        """
+        cluster_id = cluster_id or self.cluster_id
+        if not cluster_id:
+            raise AirflowException('Cluster id must be specified.')
+
+        self.log.info('Running Pyspark job {}'.format(cluster_id))
+        operation = self.client(job_service_grpc_pb.JobServiceStub).Create(
+            job_service_pb.CreateJobRequest(
+                cluster_id=cluster_id,
+                name=name,
+                pyspark_job=job_pb.PysparkJob(
+                    main_python_file_uri=main_python_file_uri,
+                    python_file_uris=python_file_uris,
+                    file_uris=file_uris,
+                    archive_uris=archive_uris,
+                    jar_file_uris=jar_file_uris,
+                    args=arguments,
+                    properties=properties,
+                )
+            )
+        )
+        return self._get_operation_result(operation, response_type=job_pb.Job)
+
+    def create_cluster(
+        self,
+        folder_id=None,
+        cluster_name=None,
+        cluster_description='',
+        cluster_image_version='1.1',
+        ssh_public_keys=None,
+        subnet_id=None,
+        services=('HDFS', 'YARN', 'MAPREDUCE', 'HIVE', 'SPARK'),
+        s3_bucket=None,
+        zone='ru-central1-b',
+        service_account_id=None,
+        masternode_resource_preset='s2.small',
+        masternode_disk_size=15,
+        masternode_disk_type='network-ssd',
+        datanode_resource_preset='s2.small',
+        datanode_disk_size=15,
+        datanode_disk_type='network-ssd',
+        datanode_count=2,
+        computenode_resource_preset='s2.small',
+        computenode_disk_size=15,
+        computenode_disk_type='network-ssd',
+        computenode_count=0,
+    ):
```

Comment:
We try to avoid so many parameters. This often results in missing params that are not passed to sdk. It seems that here users instead of passing so many arguments can just pass cluster definition?

Here's an example of Google Dataproc:
https://github.com/apache/airflow/blob/ce3635d64f2eba9616709fb9d998676aa34b0875/airflow/providers/google/cloud/operators/dataproc.py#L456-L466

More information:
https://github.com/apache/airflow/blob/master/docs/howto/custom-operator.rst

Reply:
Honestly I see no benefits in this approach. 
What authors of mentioned operator do is hiding all cluster related parameters inside kwargs,
forcing the users to go see some dicts like
https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/example_dags/example_dataproc.py#L45-L60
in various examples and docs, instead of letting them using IDE builtin autocompletion and code navigation to suggest related params.

The only good result in it, that we (yandexcloud python package maintainers) may publish new versions of the package without need to update corresponding Airflow hooks and operators,
but this will only let Airflow users to use new features without corresponding docs and examples in Airflow repo. Since we plan longterm maintenance for Airflow bindings, we are going to propose new Airflow PRs everytime we publish new features to our Yandex cloud python package, so I see no benefit of implicit passing params here.

Do you really think that proposed way with dicts passing is better for the Airflow users?
If you insist I can create some new classes for cluster configuration but imho this would be just boilerplate.

Reply:
The dict-way works for Google because that's how their Python DSK works. When you want to create a cluster you have to provide a cluster object. And it seems to be similar to Yandex (you are creating a cluster in hook before passing it to `Create`).  

In this way Airflow Operators works as described in Google Dataproc (or any other original) documentation and that's a huge advantage because:
- user can use Airflow operators in the same way he/she uses the SDK
- already available documentation works for Airflow operators
- no additional logic that has to be maintained by us (when authors no longer contribute to Airflow)
- less code

We should thrive to make Airflow hooks and operators as thin as possible. Some time ago I gave a talk about generating Airflow operators where I pointed out why it's good to go "thin":
https://youtu.be/Nr4Pp1SNXeU?t=5106

Here's a guide that we have worked in the community:
https://docs.google.com/document/d/1J1GKPWYIguxNNyeb24RxP2SxU3TXpC4UpwrvjeFLGfM/edit

Reply:
You are right. I will move as much logic as possible to yandexcloud package.

Reply:
Can you take another look, please?

Reply:
Done. Now the hook is a thin proxy object to yandexcloud package interface